sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuUserDefinedFunction.scala

```diff
@@ -89,11 +100,21 @@ trait GpuRowBasedUserDefinedFunction extends GpuExpression
   /** True if the UDF is deterministic */
   val udfDeterministic: Boolean
 
+  /** True if the UDF needs null check when converting input columns to rows */
+  def checkNull: Boolean
+
   /** The row based function of the UDF. */
   protected def evaluateRow(childrenRow: InternalRow): Any
 
   private[this] lazy val inputTypesString = children.map(_.dataType.catalogString).mkString(", ")
   private[this] lazy val outputType = dataType.catalogString
+  /**
+   * FIXME This should be called per JVM (aka process), not sure the `lazy` can help here.
+   * This is for https://github.com/NVIDIA/spark-rapids/issues/3942.
+   * And more details can be found from
+   *     https://github.com/NVIDIA/spark-rapids/pull/3997#issuecomment-957650846
+   */
+  private[this] lazy val nullSafe: Boolean = GpuUserDefinedFunction.isAssertionEnabled && checkNull
```

Comment:
This is actually called per instance of GpuUserDefinedFunction. But that is not a big deal, unless we see it impacting performance. If you really are concerned about it put change it to 

```
private[this] lazy val nullSafe: Boolean = checkNull && GpuUserDefinedFunction.isAssertionEnabled
```

That way if checkNull is false we never even try to check the assertions enabled function.

Reply:
Thanks for the suggestion, and updated this according to my latest investigaion.